### PR TITLE
Update java support to latest jdt.ls

### LINKED ISSUE #1068: 'Format Document' command isn't functional in Java Language.
Clicked the menu item 'Format Document' is no effect,When editing a java source code document.
image
But in a javascript document, this command would preformed an act of expectation.
image
And I want 'Format Document' command is functional in a **custom language**. what should l do ?

### PR DESCRIPTION
**Pull request by gorkem**
Updates the server to the latest snapshot which includes
a fix that is required for Theia. Also changes the way
language server is started to adjust to the changes
on environment variables.

fixes #1068 also improves #1332

### CODE REVIEW

File: packages/java/src/browser/java-client-contribution.ts [hunk 1]
```diff
@@ -8,44 +8,44 @@
 import { injectable, inject } from "inversify";
 import { CommandService } from "@theia/core/lib/common";
 import {
-    Window, ILanguageClient, BaseLanguageClientContribution, Workspace, Languages, LanguageClientFactory
+  Window, ILanguageClient, BaseLanguageClientContribution, Workspace, Languages, LanguageClientFactory
```
Comment (akosyakov) [inline packages/java/src/browser/java-client-contribution.ts]: Could you check your formatting setting please? It is 4 spaces for ts files.
For VS code we have it checked in: https://github.com/theia-ide/theia/blob/e41a77e57de7f5948e3c61cfe5f5f724b6e2fac1/.vscode/settings.json#L41

File: packages/java/src/node/java-contribution.ts [hunk 2]
```diff
@@ -22,59 +22,56 @@ configurations.set('linux', 'config_linux');
...
-        Promise.all([
-            this.startSocketServer(), this.startSocketServer()
-        ]).then(servers => {
-            const [inServer, outServer] = servers;
-            const inSocket = this.accept(inServer);
-            const outSocket = this.accept(outServer);
+    Promise.all(
+      [this.startSocketServer()]
+    ).then(servers => {
+      const [server] = servers;
```
Comment (kittaakos) [inline packages/java/src/node/java-contribution.ts]: Minor: perhaps, you could get rid of the array of servers and the `Promise.all` as well if we have one, right? And write 
```typescript
this.startSocketServer().then(server => {
  // blabla
});
```
instead.

File: packages/java/src/browser/java-client-contribution.ts (hunk 1 above)
Comment (gorkem) [reply packages/java/src/browser/java-client-contribution.ts]: Naturally I use Theia to develop Theia. I think it is odd that we are depending on VSCode on these.

**Comment by marcdumais-work** [comment]
@gorkem I gave this PR a quick try on Linux. I can confirm that now "Format Document" seems to work, and it did not on master.

However the diagnostics, at least on Linux, seem broken. Whatever I type, it seems that no new error markers are created ("old" ones I think were saved in the browser and restored but can't be cleared by fixing the issue). That works better on master.

Completion seems not to work as well also, with this PR.

**Comment by marcdumais-work** [comment]
>However the diagnostics, at least on Linux, seem broken. 

Actually, not completely broken - the diagnostics seem to kick-in once in a while.

File: packages/java/src/browser/java-client-contribution.ts (hunk 1 above)
Comment (akosyakov) [reply packages/java/src/browser/java-client-contribution.ts]: true, but Theia defaults is 4 spaces: https://github.com/theia-ide/theia/blob/f4de4f572bdebd7dadae5e7a3323e39fb346d566/packages/editor/src/browser/editor-preferences.ts#L24

We can check in `.theia/setting.json` as well.

**Comment by gorkem** [comment]
@marcdumais-work correct. That is because for some reason that I have not discovered yet jdt.ls is not receiveing didOpen/didChange/didClose events. Those events are needed to update the java AST model. At the moment the diagnostics are happening because of a side-effect of other events.

**Comment by kittaakos** [comment]
I have restarted the Windows CI.

**Comment by gorkem** [comment]
Fixed comments and rebased to master should be ready to go.

**Comment by marcdumais-work** [comment]
@gorkem what about completion? It still seems not to work well, with the latest version of this PR. 
peek 2018-03-29 06-54


However I re-tried on master and it's now not working at all - maybe I was lucky it worked yesterday. So overall this PR is not making things much worse, so no objection from me to merge. I hope you can continue to improve Java LS support.

**Comment by marcdumais-work** [comment]
BTW the Travis failures seemed unrelated - I have restarted that build.

**Comment by gorkem** [comment]
@marcdumais-work not working at all is not normal, I will check again. The other oddities are expected because jdt.ls is still not receiving any of th TextDocumentSync events so it is likely that the Java model is out of sync with the editing.

**Comment by marcdumais-work** [comment]
@gorkem the completions "not working at all" was when I tried on master today, not with this PR. It was working yesterday for me on master, so maybe something is flaky.

**Comment by gorkem** [comment]
finally found the culprit. jdt.ls is hitting the bug fixed with https://github.com/TypeFox/vscode-languageserver-node/commit/e0c1f194dabede0409f0d8f4e4e2eeccc65e1535. @svenefftinge When can we update the library?

**Comment by svenefftinge** [comment]
We want to bump to the latest monaco, that would include rebasing this library, as well. But it is a bit more work.

**Comment by AlexTugarev** [comment]
@svenefftinge, may I do the patch release?

File: packages/java/src/node/java-contribution.ts [hunk 3]
```diff
@@ -48,7 +48,7 @@ export class JavaContribution extends BaseLanguageServerContribution {
 
         if (DEBUG_MODE) {
             args.push(
-                '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=1044',
+                '-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=1044',
```
Comment (AlexTugarev) [inline packages/java/src/node/java-contribution.ts]: @gorkem, do we really want to suspend the Jvm? I guess it's useful to debug the LS handshake. Just asking.

**Comment by AlexTugarev** [comment]
@gorkem this looks good to me. We need those changes!

On the `vscode-base-languageclient`, I've published new version containing the bugfix. Now updating the chain in order to consume it in Theia. #1539

File: packages/java/src/node/java-contribution.ts (hunk 3 above)
Comment (gorkem) [reply packages/java/src/node/java-contribution.ts]: Good catch. It should be fixed now.

**Comment by AlexTugarev** [comment]
@gorkem, the CI build is passed, do you want to merge?

**Comment by gorkem** [comment]
OK, I will rebase and merge

**Comment by AlexTugarev** [comment]
@gorkem, not whether how this is related, but trying out latest changes shows that the java editor is broken. Looking at the errors I see at begins with trying to register `java.edit.organizeImports` command, and the error message is that's already registered. Then I see lot's of follow-up errors. Trying to complete `System.o|ut` shows sometimes errors like `INVALID range, must be single line and on the same line`. 

Suggestions seems to be off, too:

screen shot 2018-04-06 at 11 50 19

Then I see protocol error messages like:

```
"Apr 6, 2018 11:57:51 AM Failed to create linked resource from file:///Users/tugarev/Projects/web2/Main123.java to jdt.ls-java-project
Resource '/jdt.ls-java-project/src/Main123.java' already exists.
org.eclipse.core.internal.resources.ResourceException(/jdt.ls-java-project/src/Main123.java)[367]: java.lang.Exception: Resource '/jdt.ls-java-project/src/Main123.java' already exists.
	at org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:39)
	at org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:35)
	at org.eclipse.core.internal.resources.Workspace.createResource(Workspace.java:1362)
	at org.eclipse.core.internal.resources.Workspace.createResource(Workspace.java:1297)
	at org.eclipse.core.internal.resources.Resource.createLink(Resource.java:636)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.getFakeCompilationUnit(JDTUtils.java:207)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.resolveCompilationUnit(JDTUtils.java:140)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.resolveCompilationUnit(JDTUtils.java:113)
	at org.eclipse.jdt.ls.core.internal.handlers.CodeActionHandler.getCodeActionCommands(CodeActionHandler.java:59)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$12(JDTLanguageServer.java:515)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$21(JDTLanguageServer.java:703)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
"
```

Do you have any issues?

**Comment by gorkem** [comment]
@AlexTugarev I think you are hitting an issue that jdt.ls faces time to time. This usually happens when we change eclipse versions or a workspace (the eclipse one) is somewhat corrupted. Deleting the workspace should fix it

**Comment by gorkem** [comment]
Rebased and updated jdt.ls version

**Comment by AlexTugarev** [comment]
Thanks for the hint. I’ll try it.

**Comment by AlexTugarev** [comment]
@gorkem, I tried with fresh project and made sure there is no temp workspace data picked up. 

The Java editor is still not useable, and it most likely because of the missing `TextDocumentSync`. This gif shows how the LS is some changes behind the editor:


But I also still see errors like "java.edit.organizeImports" (with a lot of follow ups). I just verified this by patching `this._capabilites.textDocumentSync` to `2`  in `/vscode-base-languageclient/lib/base.js`. Also code lenses seems to disappear and reappear while typing, which is very disturbing.



Could you check the "java.edit.organizeImports" issue?

**Comment by gorkem** [comment]
'java.edit.organizeImports' looks like it is registered multiple times because jdt.ls dynamically registers it to enable 'workspace/executeCommand' calls. However it is also registered with a 'CommandContribution' at startup to enable Menu and keybinding registration. I do not think there is anything jdt.ls can do for it. LSP implementation should probably be changed to avoid re-registration.

**Comment by gorkem** [comment]
created #1684 for the multiple command registration issue.

**Comment by AlexTugarev** [comment]
@gorkem, I've created this issue for JDT LS https://github.com/eclipse/eclipse.jdt.ls/issues/629. I think this is responsible for the hiccups, which appears when typing.

**Pull request merged**